contracts/swap/proxy.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package swap package wraps the 'swap' Ethereum smart contract.
+// It is an abstraction layer to hide implementation details about the different
+// Swap contract iterations (SimpleSwap, Swap, etc.)
+package swap
+
+import (
+	"bytes"
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// Backend wraps all methods required for contract deployment.
+type Backend interface {
+	bind.ContractBackend
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	//TODO: needed? BalanceAt(ctx context.Context, address common.Address, blockNum *big.Int) (*big.Int, error)
+}
+
+// Proxy is a proxy object for Swap contracts.
+// Currently we only have SimpleSwap, but full Swap may be a different contract.
+// To abstract contract references and not have to refactor too much code
```

Comment:
According to Ralph and Rinke, we can expect all Swap implementations to retain their function signatures, so this wrapper would possibly be unnecessary; Simple Swap would be a subset of Soft Swap, and the latter a subset of Swap.

I suggest we keep the wrapper for now and discuss deprecating it later. What do you think?

Reply:
Also, after discussing it with the group over here we think it would be beneficial to–at some point, no need to do it right now–have a sort of class diagram (or similar) to understand the connection/relation between all the swap entities, like the different `Swap` structs, the `SwapService` pointer, the `swapInstance` variable, etc.

Reply:
I agree that the wrapper can be removed as soon as it is obsolete. Currently we would need a refactoring of the contract in order to do that. So let's keep it until we are stable with the code and we all agree we don't need it.